File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/playlists/PlayListsPresenter.java
package com.github.anrimian.musicplayer.ui.playlist_screens.playlists;

import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.List;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public class PlayListsPresenter extends MvpPresenter<PlayListsView> {

    private final PlayListsInteractor playListsInteractor;
    private final Scheduler uiScheduler;
    private final ErrorParser errorParser;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();

    public PlayListsPresenter(PlayListsInteractor playListsInteractor,
                              Scheduler uiScheduler,
                              ErrorParser errorParser) {
        this.playListsInteractor = playListsInteractor;
        this.uiScheduler = uiScheduler;
        this.errorParser = errorParser;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        subscribeOnPlayLists();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onPlayListLongClick(PlayList playList) {
        getViewState().showPlayListMenu(playList);
    }

    void onDeletePlayListButtonClicked(PlayList playList) {
        getViewState().showConfirmDeletePlayListDialog(playList);
    }

    void onDeletePlayListDialogConfirmed(PlayList playList) {
        playListsInteractor.deletePlayList(playList.getId())
                .observeOn(uiScheduler)
                .subscribe(() -> onPlayListDeleted(playList), this::onPlayListDeletingError);
    }

    void onFragmentMovedToTop() {
        playListsInteractor.setSelectedPlayListScreen(0);
    }

    void onChangePlayListNameButtonClicked(PlayList playList) {
        getViewState().showEditPlayListNameDialog(playList);
    }

    private void onPlayListDeletingError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showDeletePlayListError(errorCommand);
    }

    private void onPlayListDeleted(PlayList playList) {
        getViewState().showPlayListDeleteSuccess(playList);
    }

    private void subscribeOnPlayLists() {
        getViewState().showLoading();
        presenterDisposable.add(playListsInteractor.getPlayListsObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onPlayListsReceived));
    }

    private void onPlayListsReceived(List<PlayList> list) {
        getViewState().updateList(list);
        if (list.isEmpty()) {
            getViewState().showEmptyList();
        } else {
            getViewState().showList();
        }
    }
}

File: app/src/main/java/com.github.anrimian.musicplayer/ui/playlist_screens/playlists/PlayListsPresenter.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.playlists

import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Scheduler
import java.util.*

class PlayListsPresenter(
        private val playListsInteractor: PlayListsInteractor,
        uiScheduler: Scheduler,
        errorParser: ErrorParser
) : AppPresenter<PlayListsView>(uiScheduler, errorParser) {
    
    private var playLists: List<PlayList> = ArrayList()
    
    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        subscribeOnPlayLists()
    }

    fun onStop(listPosition: ListPosition?) {
        playListsInteractor.saveListPosition(listPosition)
    }

    fun onDeletePlayListButtonClicked(playList: PlayList) {
        viewState.showConfirmDeletePlayListDialog(playList)
    }

    fun onDeletePlayListDialogConfirmed(playList: PlayList) {
        playListsInteractor.deletePlayList(playList.id)
                .subscribeOnUi({ onPlayListDeleted(playList) }, this::onPlayListDeletingError)
    }

    fun onFragmentMovedToTop() {
        playListsInteractor.setSelectedPlayListScreen(0)
    }

    fun onChangePlayListNameButtonClicked(playList: PlayList) {
        viewState.showEditPlayListNameDialog(playList)
    }

    private fun onPlayListDeletingError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showDeletePlayListError(errorCommand)
    }

    private fun onPlayListDeleted(playList: PlayList) {
        viewState.showPlayListDeleteSuccess(playList)
    }

    private fun subscribeOnPlayLists() {
        viewState.showLoading()
        playListsInteractor.playListsObservable.unsafeSubscribeOnUi(this::onPlayListsReceived)
    }

    private fun onPlayListsReceived(list: List<PlayList>) {
        val firstReceive = playLists.isEmpty()
        playLists = list
        viewState.updateList(list)
        if (list.isEmpty()) {
            viewState.showEmptyList()
        } else {
            viewState.showList()
            if (firstReceive) {
                val listPosition = playListsInteractor.savedListPosition
                if (listPosition != null) {
                    viewState.restoreListPosition(listPosition)
                }
            }
        }
    }
}